artemis-protocols/artemis-mqtt-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/mqtt/MQTTSessionStateManager.java

```diff
@@ -0,0 +1,181 @@
+    * one instance of MQTTSessionStateManager per-broker with the understanding that there can be multiple brokers in
+    * the same JVM.
+    */
+   public static synchronized MQTTSessionStateManager getInstance(ActiveMQServer server) throws Exception {
+      MQTTSessionStateManager instance = INSTANCES.get(System.identityHashCode(server));
+      if (instance == null) {
+         instance = new MQTTSessionStateManager(server);
+         INSTANCES.put(System.identityHashCode(server), instance);
+      }
+
+      return instance;
+   }
+
+   public static synchronized void removeInstance(ActiveMQServer server) {
+      INSTANCES.remove(System.identityHashCode(server));
+   }
+
+   private MQTTSessionStateManager(ActiveMQServer server) throws Exception {
+      this.server = server;
+      sessionStore = server.createQueue(new QueueConfiguration(MQTTUtil.MQTT_SESSION_STORE).setRoutingType(RoutingType.ANYCAST).setLastValue(true).setDurable(true).setInternal(true).setAutoCreateAddress(true), true);
+
+      // load session data from queue
+      try (LinkedListIterator<MessageReference> iterator = sessionStore.browserIterator()) {
+         try {
+            while (iterator.hasNext()) {
+               MessageReference ref = iterator.next();
+               String clientId = ref.getMessage().getStringProperty(Message.HDR_LAST_VALUE_NAME);
+               MQTTSessionState sessionState = new MQTTSessionState((CoreMessage) ref.getMessage(), this);
+               sessionStates.put(clientId, sessionState);
+            }
+         } catch (NoSuchElementException ignored) {
+            // this could happen through paging browsing
+         }
+      }
+   }
+
+   public void scanSessions() {
+      List<String> toRemove = new ArrayList();
+      for (Map.Entry<String, MQTTSessionState> entry : sessionStates.entrySet()) {
+         MQTTSessionState state = entry.getValue();
+         logger.debug("Inspecting session: {}", state);
+         int sessionExpiryInterval = state.getClientSessionExpiryInterval();
+         if (!state.isAttached() && sessionExpiryInterval > 0 && state.getDisconnectedTime() + (sessionExpiryInterval * 1000) < System.currentTimeMillis()) {
+            toRemove.add(entry.getKey());
+         }
+         if (state.isWill() && !state.isAttached() && state.isFailed() && state.getWillDelayInterval() > 0 && state.getDisconnectedTime() + (state.getWillDelayInterval() * 1000) < System.currentTimeMillis()) {
+            state.getSession().sendWillMessage();
+         }
+      }
+
+      for (String key : toRemove) {
+         try {
+            MQTTSessionState state = removeSessionState(key);
+            if (state != null && state.isWill() && !state.isAttached() && state.isFailed()) {
+               state.getSession().sendWillMessage();
+            }
+         } catch (Exception e) {
+            // TODO: make this a real error message
+            e.printStackTrace();
```

Comment:
still TODO?

Reply:
That `TODO` was copied from `MQTTProtocolManager`. It is still to-do, but not related to this PR.

Reply:
There isnt a TODO in MQTTProtocolManager that I can see, and none was removed in the diff seemingly moving this code, so it appears to be getting introduced here along with the try-catch it is in.

Reply:
My apologies. I totally misremembered what happened here. I've resolved the issue. Thanks for verifying.